### ISSUE TITLE
pull mode should be allowed to (always?) pull parent macro input first

### ISSUE THREAD

**Issue by JNmpi** [pyiron/pyiron_workflow#54]
Thanks @liamhuber for implementing the pull mode. The examples you provided work nicely but when I try to run it for the lammps example it fails. The code for the calling cell is:

          wf = Workflow('Lammps')
          structure = wf.create.atomistics.Bulk('Al', cubic=True)
          repeat = wf.create.lammps.Repeat(structure = structure, repeat_scalar=3)
          
          lammps = wf.create.atomistic_codes.Lammps(structure=repeat, label='lammps')
          repeat.pull()  # works
          # lammps.structure.pull() # does not work
          # lammps.list_pots.pull() # does not work
          
          # lammps() # works

The lammps node in atomistic_codes is:

        @macro_node()
        def lammps(wf: Macro) -> None:
            from pyiron_contrib.tinybase.shell import ExecutablePathResolver
        
            structure = wf.create.lammps.Structure()
        
            potential = wf.create.lammps.Potential(structure=structure)
        
            wf.list_pots = wf.create.lammps.ListPotentials(structure=structure)

The rest of the node is identical to the one in the atomistic_codes library. Do you have any idea what is going on and why the above code fails. 

When writing the example I realized that it would be nice to add optional functions to a macro which are only called via an explicit `pull()`, but not if we run the workflow. A possible syntax would be:

      wf.add_function(list_potential=wf.create.lammps.ListPotentials(structure=wf.structure))

**Comment by liamhuber** [comment]
Hi @JNmpi,

This is working as intended; macros now have their own IO instead of passing their children's IO by reference; when you `.pull()` a child of a macro, it will only pull _inside_ the scope of the macro -- i.e. whatever input the macro currently has is the most up-stream data you'll get. I am still a fan of this design choice, but I agree we can give a quality-of-life shortcut here.

My version of the `Lammps` macro from the head of #33 looks a little different, but it was an easy fix to get it to have `list_pots` like yours. You can then accomplish what you want by manually pulling all the input to `lammps` and then fetching that up-to-date input:

```python
from pyiron_workflow import Workflow

wf = Workflow('Lammps')
structure = wf.create.atomistics.Bulk('Al', cubic=True)
repeat = wf.create.lammps.Repeat(structure = structure, repeat_scalar=3)

lammps = wf.create.atomistic_codes.Lammps(structure=repeat, label='lammps')
repeat.pull()  # works -- updates the output of repeat
lammps.inputs.fetch()  # gets the `repeat` output as input
# lammps.structure.pull() # works now
lammps.list_pots.pull() # works now

# lammps.pull() # worked all a long, 
# but ran the _whole_ macro, which we might not want
```

`.pull()` does this too but it _always_ executes after. I can see that it would be super helpful to have a shortcut for this input pulling _without_ executing after.

Maybe

```python
wf.lammps.pull_input()
```

or 

```python
wf.lammps.pull(input_only=True)  # Default is False
```

or even

```
wf.lammps.pull(then_run=False)  # Default is True
```


With these the example would look something like this:

```python
```python
from pyiron_workflow import Workflow

wf = Workflow('Lammps')
structure = wf.create.atomistics.Bulk('Al', cubic=True)
repeat = wf.create.lammps.Repeat(structure = structure, repeat_scalar=3)

lammps = wf.create.atomistic_codes.Lammps(structure=repeat, label='lammps')
lammps.pull_input()
lammps.list_pots.pull()
```

Which I would find to be a good balance between convenience and the very clean behaviour of macro-child-pulls not breaking outside their parent.

**Comment by liamhuber** [comment]
> When writing the example I realized that it would be nice to add optional functions to a macro which are only called via an explicit pull(), but not if we run the workflow.

I don't object to this, and write now the macro-building function that we decorate only takes the macro itself as an argument, so I can see room for this sort of feature. It introduces complexity though, so doing while keeping both the API and implementation nice and clear will take a bit of effort -- at the end of the day I'll probably split the idea off to its own issue so we don't lose track entirely and then de-prioritize it

**Comment by JNmpi** [comment]
Thanks @liamhuber for the quick reply and the suggestion. I am not perfectly happy with the proposed fixed. My aim is to make it as easy as possible for the users. Particularly, they should not need to see or know what is behind the scenes to run pyiron. In this respect I would really like to get rid of the explicit commands 

      lammps.pull_input()
      lammps.list_pots.pull()

and replace them by

     lammps.list_potentials()

like we have it in the present pyiron version. Thus, why not make the pull command going over the boundaries of a macro and pulling in also the input from outside. In fact, to me this would look like the expected behavior and more intuitive. The main motivation should be always to make usebility as intuitive and easy as possible, even if we have to sacrifice some of our design choices. In this respect, Is there anything critical that forces us to keep the pull activity inside the macro?

**Comment by liamhuber** [comment]
Of course! With the workshop coming up for you, I want to be time sensitive.

I don't see anything technical that prevents us from setting it so that a pulling on a child first pulls the input of its parent and then pulls the child. However, I am concerned about making this the default behaviour because triggering executions arbitrarily far upstream might unwittingly trigger expensive executions. 

For me this is one of the dangers we face by making everything no delayed. It works in the original pyiron because everything leading up to it had been executed on declaration. In this particular case running the upstream stuff is inexpensive and totally fine, but I am in favour of having the user explicitly ask for things to be recursively run upstream. At a bare minimum, choosing to run only inside the current macro scope needs to be at least as easy as first calling the external upstream stuff.

Your comment made me think of an even better syntax though, which I think will satisfy both of us (at least up to an argument about what the default value should be, haha):

```python
wf.lammps.list_pots.pull(pull_parent_input_first=True)
```

Which would then make the same call recursively until we got to the head of the entire workflow that has arbitrarily deeply nested macros. I feel this should default false; I'm not at all married to that particular name for the flag; I think it nicely allows us to get either of the two behaviours in a single line.

**Comment by liamhuber** [comment]
A second comment is that you want to replace `pull()` with `__call__`. The difference being that the current call also emits the ran signal and triggers downstream executions (iirc).

Off the top of my head I have no objections to such a change, seems alright!

**Comment by liamhuber** [comment]
@JNmpi, I implemented everything discussed here in #59. I do still think my reservations about the danger of node invocation executing nodes arbitrarily upstream are valid, but on reflection it's not clear to me which inconvenience will cause the greater weighted-average pain: (1) always pull upstream and sometimes run stuff that hurts (high pain, low probability), or (2) only pull upstream in the local scope unless you remember to specify a flag to break out (low pain, high probability). Of course both choices need to be (and are) available. For the default I ultimately came down on your side of always being very aggressive with our pulls -- `__call__` now pulls recursively before running. We can always switch the default if we get enough complaints!

The PR needs test and documentation fixes, but all the existing tests are running (actually surprisingly few updates were required for that) and I'm content with the interface.

**Comment by JNmpi** [comment]
Thanks for implementing these features, @liamhuber. I share your reservations, but find the present solution to make a complete pull the default behavior a good solution. It is for me the behavior I would expect as a user and if long-running nodes are called this is a bit up to the user. We may later introduce a tag for the modules that gives a hint as to whether this function can be evaluated on the fly (e.g. most/all of the nodes related to structure conversion) or whether this is a more expensive node (like VASP). For the latter case, when calling pull a waring could appear asking the user to confirm that this can be an expensive calculation.
